Define and measure duration for text notifications in milliseconds

dragunoff opened the pull request:
Notification duration should be the same regardless of game speed. Switch to using wall-clock time defined in milliseconds instead of game ticks. Also use the opportunity to rename the field to "Duration" because "RemoveTime" is not so clear.

Closes #20501

dragunoff committed:
Use an overload that already passes the second argument as true

dragunoff commented on OpenRA.Mods.Common/Widgets/TextNotificationsDisplayWidget.cs:
```diff
@@ -18,7 +18,7 @@ namespace OpenRA.Mods.Common.Widgets
 	public class TextNotificationsDisplayWidget : Widget
 	{
-		public readonly int RemoveTime = 0;
+		public readonly int Duration = 0;
```
I'm not entirely convinced this is a great name either but it seems better than "RemoveTime".

ThomasChr replied:
DisplayDuration?

ThomasChr replied:
And after watching the Video recommend from @IceReaper in Discord - https://m.youtube.com/watch?v=-J3wNP6u5YU - we even can name ist DisplayDurationMilliseconds - if that won‘t use too much Bytes :-)

dragunoff replied:
Yes, making clear what's the unit of measurement is great. `DisplayDurationMilliseconds` is a bit lengthy but it's unambiguous and self-documenting. Maybe just `DurationMilliseconds`?

abcdefg30 replied:
You could abbreviate Milliseconds to Ms if the former is too long for your taste. We use `DisplayDuration` elsewhere, so either that or `DisplayDurationMs`/`DisplayDurationMilliseconds` would fit imo.

abcdefg30 commented on mods/cnc/chrome/ingame-chat.yaml:
```diff
@@ -19,7 +19,7 @@ Container@CHAT_PANEL:
 					Width: PARENT_RIGHT
 					Height: PARENT_BOTTOM
-					RemoveTime: 250
+					Duration: 10000
```
This will need an update rule.

dragunoff replied:
Oh, I didn't know we can update chrome yaml. With `UpdateChromeNode`? Seems like it has never been used :crossed_fingers:

dragunoff commented:
Update: Added update rule.

PunkPun commented on OpenRA.Mods.Common/UpdateRules/Rules/TextNotificationsDisplayWidgetRemoveTime.cs:
```diff
@@ -0,0 +1,39 @@
+﻿#region Copyright & License Information
```
This needs to be placed in a new folder. It also isn't mentioned in the UpdatePath

dragunoff commented:
Update: Moved update rule and added to the update path.

PunkPun commented on OpenRA.Mods.Common/UpdateRules/UpdatePath.cs:
```diff
@@ -74,7 +74,6 @@ public class UpdatePath
 
 			new UpdatePath("release-20210321", new UpdateRule[]
```
```suggestion
			new UpdatePath("release-20210321", "playtest-20221203", new UpdateRule[]
```

dragunoff commented:
Update: fixed https://github.com/OpenRA/OpenRA/pull/20502#discussion_r1039018490

PunkPun reviewed (dismissed):
LGTM

dragunoff commented:
Update: changed field name to `DisplayDurationMs`.

abcdefg30 commented on OpenRA.Mods.Common/Widgets/TextNotificationsDisplayWidget.cs:
```diff
@@ -54,6 +62,8 @@ public override void Initialize(WidgetArgs args)
 			overflowDrawBounds.Y += Bounds.Height - visibleChildrenHeight;
 			overflowDrawBounds.Height = visibleChildrenHeight;
+
+			durationTicks = DisplayDurationMs / world.Timestep;
```
This doesn't adapt to changing the timestamp in replays.

dragunoff replied:
I gave this some thought and actually implemented scaling of the duration in replays. But I wonder - does the duration need to scale in replays at all? It's the moment a message appears on the screen that's important not whether it stays the same relative time on screen.

dragunoff committed:
Define and measure duration for text notifications in milliseconds

During a game notification duration should be the same regardless of
game speed. Switch to using wall-clock time defined in milliseconds 
instead of game ticks. Also use the opportunity to rename the field 
to "Duration" because "RemoveTime" is not so clear.

dragunoff commented:
Update: switch to measuring the time against `Game.RunTime` which is in milliseconds so no need to convert to ticks. This also fixes an issue where notifications won't be removed after a replay has ended.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/07847d625ef296a39832b4e8488c45830f90f5b1...a88860e919f823ee66d898a401926c0c5dc5f707)
